libecole/src/instance/capacitated-facility-location.cpp

```diff
@@ -0,0 +1,254 @@
+	// Transposing and asserting row major as we pass the pointer as an array to SCIP when creating constraints.
+	serving_vars = xt::transpose(std::move(serving_vars));
+	assert(serving_vars.layout() == xt::layout_type::row_major);
+
+	auto const [n_facilities, n_customers] = serving_vars.shape();
+	assert(facility_vars.size() == n_facilities);
+	assert(demands.size() == n_customers);
+	assert(capacities.size() == n_facilities);
+
+	for (std::size_t facility_idx = 0; facility_idx < n_facilities; ++facility_idx) {
+		auto const name = fmt::format("c_{}", facility_idx);
+		auto cons = scip::create_cons_basic_linear(
+			scip, name.c_str(), n_customers, &serving_vars(facility_idx, 0), demands.data(), -inf, 0.);
+		scip::call(SCIPaddCoefLinear, scip, cons.get(), facility_vars[facility_idx], -capacities[facility_idx]);
+		scip::call(SCIPaddCons, scip, cons.get());
+	}
+}
+
+/** Add n_customers * n_facilities constraint that tighten the LP relaxation.
+ *
+ * A facility cannot serve demand if it is not open.
+ * Constraints are relased automatically (through unique_ptr in scip::create_cons_basic_linear).
+ */
+auto add_tightening_cons(
+	SCIP* scip,
+	xt::xtensor<SCIP_VAR*, 2> const& serving_vars,
+	xt::xtensor<SCIP_VAR*, 1> const& facility_vars) -> void {
+	auto const inf = SCIPinfinity(scip);
+	auto const [n_customers, n_facilities] = serving_vars.shape();
+	assert(facility_vars.size() == n_facilities);
+
+	for (std::size_t customer_idx = 0; customer_idx < n_customers; ++customer_idx) {
+		for (std::size_t facility_idx = 0; facility_idx < n_facilities; ++facility_idx) {
+			auto const name = fmt::format("t_{}_{}", customer_idx, facility_idx);
+			auto const vars = std::array{serving_vars(customer_idx, facility_idx), facility_vars[facility_idx]};
+			auto constexpr coefs = std::array<scip::real, 2>{1., -1};
+			auto cons = scip::create_cons_basic_linear(scip, name.c_str(), vars.size(), vars.data(), coefs.data(), -inf, 0.);
+			scip::call(SCIPaddCons, scip, cons.get());
+		}
+	}
+}
+
+}  // namespace
+
+scip::Model CapacitatedFacilityLocationGenerator::generate_instance(
+	RandomEngine& random_engine,
+	CapacitatedFacilityLocationGenerator::Parameters parameters) {
+
+	// Sample 1D integers array in the given interval (xtensor lazy).
+	// We sample as integer as it is generally preferred by integer programming reseachers.
+	// The usual argument is that one can turn everything into integer with appropriate scaling (rational data).
+	auto randint = [&random_engine](std::size_t n, auto interval) {
+		return xt::random::randint({n}, interval.first, interval.second, random_engine);
+	};
+
+	// Customer demand
+	auto const demands = static_cast<xvector>(randint(parameters.n_customers, parameters.demand_interval));
+	// Facilities capacity for serving customer demand
+	auto capacities = static_cast<xvector>(randint(parameters.n_facilities, parameters.capacity_interval));
+	// Fixed costs for opening facilities
+	auto const fixed_costs = static_cast<xvector>(
```

Comment:
There seems to be a discrepancy from the original Python code here. The conversion of `fixed_costs` to an int seems to have been lost in the process. Is this intentional ?
https://github.com/ds4dm/learn2branch/blob/6af745f25c25007535e7fb5a1359b518528914e6/01_generate_instances.py#L475

Reply:
So we have `randint` then a `sqrt` makes that into a `float`, and then we cast to `int` again?
I guess we missed that last part.
What does `.astype(int)` do ? floor ? ceil ? truncation ?